src/drivers/barometer/bmp581/bmp581.cpp

```diff
@@ -0,0 +1,703 @@
+	int rslt;
+	uint8_t fifo_frame_sel;
+	uint8_t reg_data[2];
+
+	fifo_frame_sel = _interface->get_reg(BMP5_REG_FIFO_SEL_ADDR);
+	fifo_frame_sel = BMP5_GET_BITS_POS_0(fifo_frame_sel, BMP5_FIFO_FRAME_SEL);
+
+	rslt = _interface->get_reg_buf(BMP5_REG_OSR_CONFIG_ADDR, reg_data, 2);
+	uint8_t odr_reg = 0xFF;
+	if(rslt == OK) {
+		odr_reg = BMP5_GET_BITSLICE(reg_data[1], BMP5_ODR);
+	}
+
+	rslt = _interface->get_reg_buf(BMP5_REG_DSP_CONFIG_ADDR, reg_data, 2);
+	uint8_t set_iir_t_reg = 0xFF;
+	uint8_t set_iir_p_reg = 0xFF;
+	if(rslt == OK) {
+		set_iir_t_reg = BMP5_GET_BITS_POS_0(reg_data[1], BMP5_SET_IIR_TEMP);
+		set_iir_p_reg = BMP5_GET_BITSLICE(reg_data[1], BMP5_SET_IIR_PRESS);
+		result = true;
+	}
+
+   	/* As per datasheet odr should be less than 5Hz. But register value for 5Hz is less than 4Hz and so,
+     	* thus in this below condition odr is checked whether greater than 5Hz macro.
+     	*/
+	if((odr_reg > BMP5_ODR_05_HZ) && (fifo_frame_sel == BMP5_DISABLE) &&
+	   (set_iir_t_reg == BMP5_IIR_FILTER_BYPASS) && (set_iir_p_reg == BMP5_IIR_FILTER_BYPASS)) {
+		*powermode = BMP5_POWERMODE_DEEP_STANDBY;
+	}
+
+	return result;
+}
+
+/*!
+ *  @brief This API is used to get powermode of the sensor.
+ */
+bool BMP581::get_power_mode(uint8_t *power_mode)
+{
+	bool result = false;
+	uint8_t deep_dis;
+	uint8_t reg_data;
+	uint8_t powermode;
+
+	if(power_mode != NULL) {
+		reg_data = _interface->get_reg(BMP5_REG_ODR_CONFIG_ADDR);
+		powermode = BMP5_GET_BITS_POS_0(reg_data, BMP5_POWERMODE);
+		switch(powermode) {
+			case BMP5_POWERMODE_STANDBY:
+				deep_dis = BMP5_GET_BITSLICE(reg_data, BMP5_DEEP_DISABLE);
+
+				if(deep_dis == BMP5_DEEP_ENABLED) {
+					if(check_deepstandby_mode(power_mode)) {
+						result = true;
+					}
+				}
+				else {
+					*power_mode = BMP5_POWERMODE_STANDBY;
+					result = true;
+				}
+				break;
+                	case BMP5_POWERMODE_NORMAL:
+                    		*power_mode = BMP5_POWERMODE_NORMAL;
+				result = true;
+                    		break;
+               		 case BMP5_POWERMODE_FORCED:
+                    		*power_mode = BMP5_POWERMODE_FORCED;
+				result = true;
+                    		break;
+                	case BMP5_POWERMODE_CONTINOUS:
+                   		*power_mode = BMP5_POWERMODE_CONTINOUS;
+				result = true;
+                    		break;
+		}
+	}
+
+	return result;
+}
+/*!
+ *  @brief This API is used to set powermode of the sensor.
+ */
+bool BMP581::set_power_mode(uint8_t power_mode)
+{
+	bool result = false;
+	uint8_t lst_pwrmode;
+	uint8_t reg_data;
+	int rslt;
+
+	if(get_power_mode(&lst_pwrmode)) {
+		if(lst_pwrmode != BMP5_POWERMODE_STANDBY) {
+			reg_data = _interface->get_reg(BMP5_REG_ODR_CONFIG_ADDR);
+			reg_data = BMP5_SET_BITSLICE(reg_data, BMP5_DEEP_DISABLE, BMP5_DEEP_DISABLED);
+			reg_data = BMP5_SET_BITS_POS_0(reg_data, BMP5_POWERMODE, BMP5_POWERMODE_STANDBY);
+			rslt = _interface->set_reg(reg_data, BMP5_REG_ODR_CONFIG_ADDR);
+			if(rslt == OK) {
+				usleep(BMP5_DELAY_US_STANDBY);
```

Comment:
you can't sleep in a work queue, the queue is shared with other work items and sleeping stalls the entire queue. If you need to sleep due to timing of the chip you need to reschedule. Typically this is done with a state machine, for example see **ADIS16507.cpp**